File: win/tests/PeerResolverTests.cs
using System.Collections.Generic;

namespace LanClip.Tests
{
    // Зеркало mac/Tests/LanClipCoreTests/PeerResolverTests.swift (задача 10/21).
    static class PeerResolverTests
    {
        // Подставной IHealthProber, считающий обращения по каждому адресу — тесты
        // резолвера проверяют не только итоговый результат, но и то, что кеш не даёт
        // резолверу ходить в сеть повторно, пока не сброшен.
        class FakeProber : IHealthProber
        {
            readonly Dictionary<string, ProbeOutcome> outcomes = new Dictionary<string, ProbeOutcome>();
            public readonly Dictionary<string, int> CallCounts = new Dictionary<string, int>();

            public FakeProber(Dictionary<string, bool> alive)
            {
                foreach (KeyValuePair<string, bool> pair in alive)
                {
                    outcomes[pair.Key] = pair.Value ? ProbeOutcome.Alive : ProbeOutcome.Unreachable;
                }
            }

            public int TotalCalls
            {
                get
                {
                    int total = 0;
                    foreach (int c in CallCounts.Values) { total += c; }
                    return total;
                }
            }

            public void SetAlive(bool isAlive, string host)
            {
                outcomes[host] = isAlive ? ProbeOutcome.Alive : ProbeOutcome.Unreachable;
            }

            // Находка I10: позволяет тестам смоделировать "сосед жив, но отверг
            // токен" — третий исход, который раньше IHealthProber не мог выразить.
            public void SetOutcome(string host, ProbeOutcome outcome)
            {
                outcomes[host] = outcome;
            }

            public ProbeOutcome Probe(string host, int port, string token, int timeoutMs)
            {
                int count;
                CallCounts.TryGetValue(host, out count);
                CallCounts[host] = count + 1;

                ProbeOutcome result;
                return outcomes.TryGetValue(host, out result) ? result : ProbeOutcome.Unreachable;
            }
        }

        static Config MakeConfig(List<string> peers)
        {
            Config config = new Config();
            config.Port = Config.DefaultPort;
            config.Token = "test-token";
            config.Peers = peers;
            config.MaxBytes = Config.DefaultMaxBytes;
            config.AutoPaste = true;
            return config;
        }

        static Dictionary<string, bool> Alive(params object[] pairs)
        {
            Dictionary<string, bool> map = new Dictionary<string, bool>();
            for (int i = 0; i < pairs.Length; i += 2)
            {
                map[(string)pairs[i]] = (bool)pairs[i + 1];
            }
            return map;
        }

        public static void Register()
        {
            T.Run("resolve returns first live address", delegate
            {
                FakeProber prober = new FakeProber(Alive("10.0.0.1", true, "10.0.0.2", true));
                PeerResolver resolver = new PeerResolver(
                    MakeConfig(new List<string> { "10.0.0.1", "10.0.0.2" }), prober);

                T.Eq("10.0.0.1", resolver.Resolve(), "resolved");
            });

            T.Run("resolve skips dead addresses", delegate
            {
                FakeProber prober = new FakeProber(Alive("10.0.0.1", false, "10.0.0.2", true));
                PeerResolver resolver = new PeerResolver(
                    MakeConfig(new List<string> { "10.0.0.1", "10.0.0.2" }), prober);

                T.Eq("10.0.0.2", resolver.Resolve(), "resolved");
            });

            T.Run("resolve returns null when none alive", delegate
            {
                FakeProber prober = new FakeProber(Alive("10.0.0.1", false, "10.0.0.2", false));
                PeerResolver resolver = new PeerResolver(
                    MakeConfig(new List<string> { "10.0.0.1", "10.0.0.2" }), prober);

                T.Eq(null, resolver.Resolve(), "resolved");
            });

            T.Run("repeated resolve does not probe network again", delegate
            {
                FakeProber prober = new FakeProber(Alive("10.0.0.1", true, "10.0.0.2", true));
                PeerResolver resolver = new PeerResolver(
                    MakeConfig(new List<string> { "10.0.0.1", "10.0.0.2" }), prober);

                T.Eq("10.0.0.1", resolver.Resolve(), "first resolve");
                int callsAfterFirst = prober.TotalCalls;
                T.Eq("10.0.0.1", resolver.Resolve(), "second resolve");
                T.Eq("10.0.0.1", resolver.Resolve(), "third resolve");

                T.Eq(callsAfterFirst, prober.TotalCalls, "кеш должен предотвращать повторные обращения к сети");
            });

            T.Run("invalidate causes reprobe", delegate
            {
                FakeProber prober = new FakeProber(Alive("10.0.0.1", true));
                PeerResolver resolver = new PeerResolver(MakeConfig(new List<string> { "10.0.0.1" }), prober);

                T.Eq("10.0.0.1", resolver.Resolve(), "first resolve");
                int callsBeforeInvalidate = prober.CallCounts["10.0.0.1"];

                resolver.Invalidate();
                T.Eq("10.0.0.1", resolver.Resolve(), "resolve after invalidate");

                T.Eq(callsBeforeInvalidate + 1, prober.CallCounts["10.0.0.1"],
                    "после Invalidate() резолвер обязан снова опросить сеть");
            });

            T.Run("after invalidate former live address that died is skipped for next alive", delegate
            {
                FakeProber prober = new FakeProber(Alive("10.0.0.1", true, "10.0.0.2", true));
                PeerResolver resolver = new PeerResolver(
                    MakeConfig(new List<string> { "10.0.0.1", "10.0.0.2" }), prober);

                T.Eq("10.0.0.1", resolver.Resolve(), "first resolve");

                prober.SetAlive(false, "10.0.0.1");
                resolver.Invalidate();

                T.Eq("10.0.0.2", resolver.Resolve(), "resolve after death of first");
            });

            T.Run("invalidate restarts probing from the beginning of the list", delegate
            {
                // После Invalidate() перебор обязан идти с начала списка Peers, а не
                // продолжаться с адреса, на котором резолвер остановился в прошлый раз.
                FakeProber prober = new FakeProber(Alive("10.0.0.1", false, "10.0.0.2", true));
                PeerResolver resolver = new PeerResolver(
                    MakeConfig(new List<string> { "10.0.0.1", "10.0.0.2" }), prober);

                T.Eq("10.0.0.2", resolver.Resolve(), "first resolve");
                T.Eq(1, prober.CallCounts["10.0.0.1"], "first probe count");

                resolver.Invalidate();
                prober.SetAlive(true, "10.0.0.1");

                T.Eq("10.0.0.1", resolver.Resolve(), "resolve after invalidate");
                T.Eq(2, prober.CallCounts["10.0.0.1"], "перебор после Invalidate() должен снова начаться с 10.0.0.1");
            });

            T.Run("resolve with empty peers returns null without probing", delegate
            {
                FakeProber prober = new FakeProber(Alive());
                PeerResolver resolver = new PeerResolver(MakeConfig(new List<string>()), prober);

                T.Eq(null, resolver.Resolve(), "resolved");
                T.Eq(0, prober.TotalCalls, "no probing for empty peers");
            });

            // MARK: - I10: неверный токен неотличим от выключенного соседа

            T.Run("resolve remembers token rejection when no peer is alive", delegate
            {
                FakeProber prober = new FakeProber(Alive());
                prober.SetOutcome("10.0.0.1", ProbeOutcome.RejectedToken);
                prober.SetOutcome("10.0.0.2", ProbeOutcome.Unreachable);
                PeerResolver resolver = new PeerResolver(
                    MakeConfig(new List<string> { "10.0.0.1", "10.0.0.2" }), prober);

                T.Eq(null, resolver.Resolve(), "resolved");
                T.True(resolver.LastResolveSawTokenRejection,
                    "хотя бы один сосед ответил 401 — это обязано отличаться от полностью мёртвой сети");
            });

            T.Run("resolve does not report token rejection when all peers are unreachable", delegate
            {
                FakeProber prober = new FakeProber(Alive());
                prober.SetOutcome("10.0.0.1", ProbeOutcome.Unreachable);
                prober.SetOutcome("10.0.0.2", ProbeOutcome.Unreachable);
                PeerResolver resolver = new PeerResolver(
                    MakeConfig(new List<string> { "10.0.0.1", "10.0.0.2" }), prober);

                T.Eq(null, resolver.Resolve(), "resolved");
                T.True(!resolver.LastResolveSawTokenRejection,
                    "ни один сосед не ответил вовсе — это не то же самое, что отвергнутый токен");
            });

            T.Run("successful resolve clears prior token rejection flag", delegate
            {
                FakeProber prober = new FakeProber(Alive());
                prober.SetOutcome("10.0.0.1", ProbeOutcome.RejectedToken);
                PeerResolver resolver = new PeerResolver(MakeConfig(new List<string> { "10.0.0.1" }), prober);

                T.Eq(null, resolver.Resolve(), "resolved");
                T.True(resolver.LastResolveSawTokenRejection, "premise");

                prober.SetOutcome("10.0.0.1", ProbeOutcome.Alive);
                resolver.Invalidate();
                T.Eq("10.0.0.1", resolver.Resolve(), "resolved after fix");
                T.True(!resolver.LastResolveSawTokenRejection,
                    "успешный Resolve() обязан сбросить прежний флаг отверженного токена");
            });
        }
    }
}

File: mac/Tests/LanClipCoreTests/PeerResolverTests.swift
import XCTest
@testable import LanClipCore

/// Подставной `HealthProbing`, считающий обращения по каждому адресу — тесты
/// резолвера проверяют не только итоговый результат, но и то, что кеш не даёт
/// резолверу ходить в сеть повторно, пока не сброшен.
private final class FakeProber: HealthProbing, @unchecked Sendable {
    private let lock = NSLock()
    private var outcomes: [String: ProbeOutcome]
    private(set) var callCounts: [String: Int] = [:]

    init(alive: [String: Bool]) {
        self.outcomes = alive.mapValues { $0 ? .alive : .unreachable }
    }

    var totalCalls: Int {
        lock.lock()
        defer { lock.unlock() }
        return callCounts.values.reduce(0, +)
    }

    func setAlive(_ isAlive: Bool, for host: String) {
        lock.lock()
        outcomes[host] = isAlive ? .alive : .unreachable
        lock.unlock()
    }

    /// Находка I10: позволяет тестам смоделировать "сосед жив, но отверг токен" —
    /// третий исход, который раньше `HealthProbing` не мог выразить вовсе.
    func setOutcome(_ outcome: ProbeOutcome, for host: String) {
        lock.lock()
        outcomes[host] = outcome
        lock.unlock()
    }

    func probe(host: String, port: Int, token: String, timeout: TimeInterval) -> ProbeOutcome {
        lock.lock()
        callCounts[host, default: 0] += 1
        let result = outcomes[host] ?? .unreachable
        lock.unlock()
        return result
    }
}

final class PeerResolverTests: XCTestCase {
    private func makeConfig(peers: [String]) -> Config {
        Config(token: "test-token", peers: peers)
    }

    func testResolveReturnsFirstLiveAddress() {
        let prober = FakeProber(alive: ["10.0.0.1": true, "10.0.0.2": true])
        let resolver = PeerResolver(config: makeConfig(peers: ["10.0.0.1", "10.0.0.2"]), prober: prober)

        XCTAssertEqual(resolver.resolve(), "10.0.0.1")
    }

    func testResolveSkipsDeadAddresses() {
        let prober = FakeProber(alive: ["10.0.0.1": false, "10.0.0.2": true])
        let resolver = PeerResolver(config: makeConfig(peers: ["10.0.0.1", "10.0.0.2"]), prober: prober)

        XCTAssertEqual(resolver.resolve(), "10.0.0.2")
    }

    func testResolveReturnsNilWhenNoneAlive() {
        let prober = FakeProber(alive: ["10.0.0.1": false, "10.0.0.2": false])
        let resolver = PeerResolver(config: makeConfig(peers: ["10.0.0.1", "10.0.0.2"]), prober: prober)

        XCTAssertNil(resolver.resolve())
    }

    func testRepeatedResolveDoesNotProbeNetworkAgain() {
        let prober = FakeProber(alive: ["10.0.0.1": true, "10.0.0.2": true])
        let resolver = PeerResolver(config: makeConfig(peers: ["10.0.0.1", "10.0.0.2"]), prober: prober)

        XCTAssertEqual(resolver.resolve(), "10.0.0.1")
        let callsAfterFirst = prober.totalCalls
        XCTAssertEqual(resolver.resolve(), "10.0.0.1")
        XCTAssertEqual(resolver.resolve(), "10.0.0.1")

        XCTAssertEqual(prober.totalCalls, callsAfterFirst, "кеш должен предотвращать повторные обращения к сети")
    }

    func testInvalidateCausesReprobe() {
        let prober = FakeProber(alive: ["10.0.0.1": true])
        let resolver = PeerResolver(config: makeConfig(peers: ["10.0.0.1"]), prober: prober)

        XCTAssertEqual(resolver.resolve(), "10.0.0.1")
        let callsBeforeInvalidate = prober.callCounts["10.0.0.1"]

        resolver.invalidate()
        XCTAssertEqual(resolver.resolve(), "10.0.0.1")

        XCTAssertEqual(prober.callCounts["10.0.0.1"], (callsBeforeInvalidate ?? 0) + 1,
                        "после invalidate() резолвер обязан снова опросить сеть")
    }

    func testAfterInvalidateFormerLiveAddressThatDiedIsSkippedForNextAlive() {
        let prober = FakeProber(alive: ["10.0.0.1": true, "10.0.0.2": true])
        let resolver = PeerResolver(config: makeConfig(peers: ["10.0.0.1", "10.0.0.2"]), prober: prober)

        XCTAssertEqual(resolver.resolve(), "10.0.0.1")

        prober.setAlive(false, for: "10.0.0.1")
        resolver.invalidate()

        XCTAssertEqual(resolver.resolve(), "10.0.0.2")
    }

    func testInvalidateRestartsProbingFromTheBeginningOfTheList() {
        // После invalidate() перебор обязан идти с начала списка peers, а не
        // продолжаться с адреса, на котором резолвер остановился в прошлый раз.
        let prober = FakeProber(alive: ["10.0.0.1": false, "10.0.0.2": true])
        let resolver = PeerResolver(config: makeConfig(peers: ["10.0.0.1", "10.0.0.2"]), prober: prober)

        XCTAssertEqual(resolver.resolve(), "10.0.0.2")
        XCTAssertEqual(prober.callCounts["10.0.0.1"], 1)

        resolver.invalidate()
        prober.setAlive(true, for: "10.0.0.1")

        XCTAssertEqual(resolver.resolve(), "10.0.0.1")
        XCTAssertEqual(prober.callCounts["10.0.0.1"], 2, "перебор после invalidate() должен снова начаться с 10.0.0.1")
    }

    func testResolveWithEmptyPeersReturnsNilWithoutProbing() {
        let prober = FakeProber(alive: [:])
        let resolver = PeerResolver(config: makeConfig(peers: []), prober: prober)

        XCTAssertNil(resolver.resolve())
        XCTAssertEqual(prober.totalCalls, 0)
    }

    // MARK: - I10: неверный токен неотличим от выключенного соседа

    func testResolveRemembersTokenRejectionWhenNoPeerIsAlive() {
        let prober = FakeProber(alive: [:])
        prober.setOutcome(.rejectedToken, for: "10.0.0.1")
        prober.setOutcome(.unreachable, for: "10.0.0.2")
        let resolver = PeerResolver(config: makeConfig(peers: ["10.0.0.1", "10.0.0.2"]), prober: prober)

        XCTAssertNil(resolver.resolve())
        XCTAssertTrue(resolver.lastResolveSawTokenRejection,
                       "хотя бы один сосед ответил 401 — это обязано отличаться от полностью мёртвой сети")
    }

    func testResolveDoesNotReportTokenRejectionWhenAllPeersAreUnreachable() {
        let prober = FakeProber(alive: [:])
        prober.setOutcome(.unreachable, for: "10.0.0.1")
        prober.setOutcome(.unreachable, for: "10.0.0.2")
        let resolver = PeerResolver(config: makeConfig(peers: ["10.0.0.1", "10.0.0.2"]), prober: prober)

        XCTAssertNil(resolver.resolve())
        XCTAssertFalse(resolver.lastResolveSawTokenRejection,
                        "ни один сосед не ответил вовсе — это не то же самое, что отвергнутый токен")
    }

    func testSuccessfulResolveClearsPriorTokenRejectionFlag() {
        let prober = FakeProber(alive: [:])
        prober.setOutcome(.rejectedToken, for: "10.0.0.1")
        let resolver = PeerResolver(config: makeConfig(peers: ["10.0.0.1"]), prober: prober)
        XCTAssertNil(resolver.resolve())
        XCTAssertTrue(resolver.lastResolveSawTokenRejection)

        prober.setOutcome(.alive, for: "10.0.0.1")
        resolver.invalidate()
        XCTAssertEqual(resolver.resolve(), "10.0.0.1")
        XCTAssertFalse(resolver.lastResolveSawTokenRejection,
                        "успешный resolve() обязан сбросить прежний флаг отверженного токена")
    }
}
